### PR TITLE
Added ParseException

### PR DESCRIPTION
**Pull request by AndyG**
Added an exception class that will help debug issues in the parser against certain inputs as they arise.

### CODE REVIEW

File: simpleast-core/src/main/java/com/discord/simpleast/core/parser/Parser.kt [hunk 1]
```diff
@@ -82,7 +82,11 @@ open class Parser<R, T : Node<R>> @JvmOverloads constructor(private val enableDe
             remainingParses.push(newBuilder)
           }
 
-          lastCapture = matcher.group(0)
+          try {
+            lastCapture = matcher.group(0)
+          } catch (throwable: Throwable) {
+            throw ParseException(message = "matcher found no matches", source = source, cause = throwable)
```
Comment (lytefast) [inline simpleast-core/src/main/java/com/discord/simpleast/core/parser/Parser.kt]: nit: not sure you need named params here.

File: simpleast-core/src/main/java/com/discord/simpleast/core/parser/Parser.kt [hunk 2]
```diff
@@ -92,7 +96,7 @@ open class Parser<R, T : Node<R>> @JvmOverloads constructor(private val enableDe
       }
 
       if (!foundRule) {
-        throw RuntimeException("failed to find rule to match source: \"$inspectionSource\"")
+        throw ParseException("failed to find rule to match source", source)
```
Comment (lytefast) [inline simpleast-core/src/main/java/com/discord/simpleast/core/parser/Parser.kt]: 1. docs need to be updated to say this is being thrown.
2. I would also update the sample/test code to include handling for this error.
3. Update README?

Comment (AndyG) [reply simpleast-core/src/main/java/com/discord/simpleast/core/parser/Parser.kt]: I don't wanna add it to the README given that this isn't an expected case for users to handle, but I can at least mention it in the doc and have a test.

Comment (AndyG) [reply simpleast-core/src/main/java/com/discord/simpleast/core/parser/Parser.kt]: actually I can't test for it... since we don't know the cause...

**Pull request merged**